### ISSUE TITLE
Memory leak on apache

### ISSUE THREAD

**Issue by Marcool04** [SpiderLabs/ModSecurity#1947]
**Describe the bug**

I am seeing a memory leak on a Raspberrypi running archlinuxarm, on an Apache webserver.

**Logs and dumps**
Here is the clearest illustration of the leak:
![screenshot 2018-11-04 at 14 51 26](https://user-images.githubusercontent.com/15948152/47964992-62ce8d80-e041-11e8-9176-f35181dd3c4d.png)
Server is started at 14:23:00 with mod_security included. Then restarted a little after 14:37:00 with it deactivated. Open webpage (netdata) is the only client side activity producing requests in both cases.

Audit log:
[blank]

Debug log: 
https://pastebin.com/L2Ugiv0w

Apache access log + error log (joined and sent to journald):
https://pastebin.com/Zf2ibirV

**To Reproduce**
The server is idle except for the netdata instance being served over a reverse proxy, which I am using to monitor ram usage.

**Expected behavior**
No memory leak :)

**Server**
 - ModSecurity version (and connector):
->Modsecurity 3.0.2-3 (installed from [libmodsecurity](https://archlinuxarm.org/packages/armv7h/libmodsecurity) package in archlinuxarm repository)
````
$ pacman -Qi libmodsecurity
Repository      : community
Name            : libmodsecurity
Version         : 1:3.0.2-3
Description     : ModSecurity v3 library component
Architecture    : armv7h
URL             : https://github.com/SpiderLabs/ModSecurity/tree/v3/master
````
->Modsecurity-apache compiled locally from a github pull on master branch (with HEAD at [96fbca4](https://github.com/SpiderLabs/ModSecurity-apache/commit/96fbca43a49c977d12f2a86e2a5ba722b04f0597)) (see also "Context" not below).

 - WebServer:
Apache 2.4.37-1
````
$ pacman -Qi apache
Repository      : extra
Name            : apache
Version         : 2.4.37-1
Description     : A high performance Unix-based HTTP server
Architecture    : armv7h
URL             : http://www.apache.org/dist/httpd
````

 - OS (and distro): 
````
$ uname -a
Linux [HOST] 4.14.78-2-ARCH #1 SMP Sun Oct 28 17:05:07 UTC 2018 armv7l GNU/Linux
````

**Rule Set (please complete the following information):**
OWASP core rule set is installed from SpiderLabs github release [3.0.2](https://github.com/SpiderLabs/owasp-modsecurity-crs/releases/tag/v3.0.2)

**Additional context**
When building the modsecurity-apache connector, I had to apply the following change:
````
diff --git a/src/mod_security3.c b/src/mod_security3.c
index 6198204..b55d33b 100644
--- a/src/mod_security3.c
+++ b/src/mod_security3.c
@@ -134,7 +134,7 @@ static void store_tx_context(msc_t *msr, request_rec *r)
 static msc_t *create_tx_context(request_rec *r) {
     msc_t *msr = NULL;
     msc_conf_t *z = NULL;
-    char *unique_id = NULL;
+    /*char *unique_id = NULL;*/
 
     z = (msc_conf_t *)ap_get_module_config(r->per_dir_config,
             &security3_module);
@@ -145,14 +145,14 @@ static msc_t *create_tx_context(request_rec *r) {
     }
 
     msr->r = r;
-    unique_id = getenv("UNIQUE_ID");
+    /*unique_id = getenv("UNIQUE_ID");
     if (unique_id != NULL || strlen(unique_id) > 0) {
         msr->t = msc_new_transaction_with_id(msc_apache->modsec,
             (Rules *)z->rules_set, unique_id, (void *)r);
-    } else {
+    } else {*/
         msr->t = msc_new_transaction(msc_apache->modsec,
             (Rules *)z->rules_set, (void *)r);
-    }
+    /*}*/
 
     store_tx_context(msr, r);
 ````
I was otherwise met with an error stating the symbol "msc_new_transaction_with_id" was undefined upon loading modsecurity into apache (the compiler was also giving a warning that the symbol was used but not defined).
Not sure that's of any relevance…

**Comment by victorhora** [comment]
Hi @Marcool04 

Thanks for your comprehensive report :)

From what I can tell, the [libModSecurity package for ArchLinuxARM](https://archlinuxarm.org/packages/aarch64/libmodsecurity/files/PKGBUILD) is built on top of 3.0.2 released version. 

I believe the issue you are reporting should not happen on the current codebase (pre-3.0.3) from v3/master. There are significant [changes and improvements](https://github.com/SpiderLabs/ModSecurity/blob/v3/master/CHANGES) between the two versions and being so I recommend upgrading it.

As for the issue with the apache_connector, this was due to the introduction of the transaction constructor at https://github.com/SpiderLabs/ModSecurity/commit/4dd28127577eca0bf854db5c134a4134db6ff1ec.
Commit https://github.com/SpiderLabs/ModSecurity-apache/commit/8bb515e93ddf0209f196fef2fa5f0a255e34de24 was added in order to make usage of the new library feature.

As of now this can be solved by either upgrading libModSecurity or removing commit https://github.com/SpiderLabs/ModSecurity-apache/commit/8bb515e93ddf0209f196fef2fa5f0a255e34de24 from your apache-connector branch.

**Comment by Marcool04** [comment]
Hi @victorhora 
Thanks for getting back to me so fast.
Indeed, I see a number of memory leek fixes in the changelog from 3.0.2 to 3.0.3. As the arch linux policy is to stick to upstream (stable) as close as possible, I would imagine that the upgrade will happen only when v3.0.3 is released (oh, I see it just was 2 hours ago, well, lets give the libmodsecurity package maintainer a while to get to it! :) )
As for the connector issue with msc_new_transaction_with_id I had found that commit using git blame and that's basically how I knew what to fix in order to get it to compile and run. But now the new release is out I'll try without the fix.
I'll make sure to check what the new version yields on my setup. Sorry for posting an issue on an outdated release (or a close to be outdated one anyhow).
Thanks for taking the time to read and answer. Keep up the good work!
All the best,
Mark.
